### PR TITLE
manual blur

### LINKED ISSUE #2: new feature: manually add blur
**main feature for v0.2 release**

after AI returns blur positions, instead of just having a final image to download, render the image to a Canvas and turn the AI blur positions into draggable rectangles. Add ability to create new rectangles and arrange them manually.

Potentially helpful reference for draggables: https://stackoverflow.com/questions/18034637/how-to-make-html5-draggable-objects-over-canvas

Not super sure how to implement blur atm

The final download can be a `context.getImageData()` call so should be pretty straightforward

### PR DESCRIPTION
**Pull request by ashlsun**
remaining issues:
- ~the first output image is always squished (??)~ @wwsalmon: fixed, `detectFaces` was applying `resize` to `image` input, so the squished version was passed to `blurFaces` later. just need to clone the JIMP object properly
-  ~when resizing the window, the DOM boxes don't move/resize accordingly~  fixed :D
- changing the sliders still affect how blur gets applied from editing mode
  - "Save edits" should ignore changes to padding and blur
  - clicking "Blur" again should close editing mode

### CODE REVIEW

File: p3-onnx-full/public/controller.js [hunk 1]
```diff
@@ -136,4 +158,226 @@ sensitivitySlider.onchange = e => {
...
+        newBox.style.height = ah + "px";
+        newBox.style.left = ax1 + "px";
+        newBox.style.top = ay1 + "px";
+        outputPreview.appendChild(newBox);
+        newBox = null;
+    }
+}
+
+// converts DOM element boxes into positions
+function saveBoxPositions() {
```
Comment (ashlsun) [inline p3-onnx-full/public/controller.js:274]: could be optimized by using a Set instead of an array for box positions

**Comment by wwsalmon** [comment]
issue: onresize re-adds the padding, so it causes the boxes to grow bigger with each resize

maybe theres a smarter way to do padding? for example, show it as a 10% CSS outline instead of changing the actual JS?

also -- maybe instead of ignoring the padding slider, maybe that slider can update the box sizes

**Comment by wwsalmon** [comment]
uh i think one of my changes made it so that the Blur button is unpressable after editing

**Comment by ashlsun** [comment]
ooh looks like boxes can be resized beyond the bounds of the image

**Comment by ashlsun** [comment]
> issue: onresize re-adds the padding, so it causes the boxes to grow bigger with each resize
> 
> maybe theres a smarter way to do padding? for example, show it as a 10% CSS outline instead of changing the actual JS?
> 
> also -- maybe instead of ignoring the padding slider, maybe that slider can update the box sizes

interestinggg ok! i dont think its onResize — I'm noticing that the padding only gets re-added if any boxes were edited, added, or deleted. 

I like this CSS outline idea, tho I think borders or outlines in CSS require the vertical and horizontal spacing to be the same.. maybe can find a workaround with padding https://stackoverflow.com/questions/14628601/can-i-add-background-color-only-for-padding or a wrapper element https://stackoverflow.com/questions/13474754/how-to-set-borders-thickness-in-percentages

**Comment by ashlsun** [comment]
ohhh hmmm, also, if the only edit was a resize, then the edit doesn't get saved. it seems like the little resize box prevents mouse events from reaching the parent element's mouse event listeners, which save the box positions. im looking into how to detect when this happens, maybe using [ResizeObserver](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver)

**Comment by ashlsun** [comment]
ok just tried out a CSS padding approach (which avoids some of the messiness of wrapper element) & rn it's looking rather complex, if we want to keep things consistent & have padding be a percentage of the width / height. 

Screenshot 2024-01-26 at 12 05 39 AM

keeping padding as a percentage means that when an element gets resized, the padding should also change, which is tough bc of the uncaught resize mouse events, and perhaps confusing for the user experience bc of the resize control box's placement at corner of the padding. when a user tries to resize, extra padding may cause the box to grow or shrink beyond the point the user drags to. 

i think we should either:
- set padding to be in terms of pixels instead of percentage
- have the boxes absorb their padding after every save (new box size = old box size + padding, padding resets to 0)
- disable padding / sliders on edit

**Pull request merged**